### DISCUSSION TITLE
The default_factory in response_model did not take effect

### DISCUSSION THREAD

**Discussion by bigfoxtail** [fastapi/fastapi#12985] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel, Field

app = FastAPI()


class Test(BaseModel):
    test: str


class Schema(BaseModel):
    content: Test | None = Field(default_factory=dict)


@app.get("/test", response_model=Schema)
def test():
    return {'content': None}
```

### Description

expect: {  "content": {}}

actually: {  "content": null}

### Operating System

Windows

### Operating System Details

Thanks
When a nested model is None, I want the final JSON to return an empty object. However, the returned JSON value is null. 
How can I make the JSON return an empty object instead?
I'm not sure if it's caused by Pydantic, SQLModel, or another module. I want to specify the value returned when a Model is None.

### FastAPI Version

0.115.4

### Pydantic Version

2.9.2

### Python Version

Python 3.12.7

**Comment by luzzodev** [comment]
There is a small in error in your code: default and default_factory are used only if the field is unset but in your case you are passing None that is a valid Value.

``` python
class Schema(BaseModel):
    content: Test = Field(default_factory=dict)
```

Note: you are using a default factory that differ from declared type. Default and default_factory do not trigger validation unless you explicit it

**Reply by bigfoxtail** [reply]
> There is a small in error in your code: default and default_factory are used only if the field is unset but in your case you are passing None that is a valid Value.
> 
> ```python
> class Schema(BaseModel):
>     content: Test = Field(default_factory=dict)
> ```
> 
> Note: you are using a default factory that differ from declared type. Default and default_factory do not trigger validation unless you explicit it

thank you!
How can I achieve the desired result? I've tried several methods, but the generated JSON is not the expected result { "content": null }.

**Reply by luzzodev** [reply]
``` python
class Test(BaseModel):
    test: str

class Schema(BaseModel):
    content: Test | None = Field(default=None)

    def model_dump(self, *args, **kwargs):
        dumped_model = super().model_dump(*args, **kwargs)
        if 'content' in dumped_model and dumped_model['content'] is None:
            dumped_model['content'] = {}
        return dumped_model
```

**Comment by estebanx64** [comment, marked as answer]
Hi @bigfoxtail 👋

This is a good approach to doing what you want. 🤓

```python
from fastapi import FastAPI
from pydantic import BaseModel, field_serializer
import uvicorn
from typing import Any

app = FastAPI()


class Test(BaseModel):
    test: str | None = None


class Schema(BaseModel):
    content: Test | None = None


class SchemaOut(BaseModel):
    content: Test | None = None

    @field_serializer("content", mode="plain")
    def serialize_content(self, value):
        # If the field is None, replace it with an empty dictionary
        return {} if value is None else value


@app.get("/test", response_model=SchemaOut)
def test() -> Any:
    return {"content": None}
```

Some things to mention regarding your example.
* The default factory expects a Callable-like function, so the implementation of `default_factory=dict` it's wrong
* You are typing the `Schema` model as `Test | None` which setting a `default=dict` it's wrong too.
* Finally I added another model called `SchemaOut` because I overwrite the `serialize_content` function to achieve what you want, but better having in another model just as a model to response, because if you want to use the base `Schema` model as an input, or table or do another validation within your logic we want to make sure the `serialize_content` it's not going to transform your values in any other context.

I hope this can be useful for you! 🙇

**Reply by bigfoxtail** [reply]
Thanks
This has been a great help to me; I’ve been struggling with it for several days.
